src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -16,36 +16,140 @@
 //! graph of `Unit`s, which capture these properties.
 
 use crate::core::compiler::Unit;
-use crate::core::compiler::{BuildContext, CompileMode, Context, Kind};
+use crate::core::compiler::{BuildContext, CompileMode, Kind};
 use crate::core::dependency::Kind as DepKind;
 use crate::core::package::Downloads;
-use crate::core::profiles::UnitFor;
-use crate::core::{Package, PackageId, Target};
+use crate::core::profiles::{Profile, UnitFor};
+use crate::core::resolver::Resolve;
+use crate::core::{InternedString, Package, PackageId, Target};
 use crate::CargoResult;
 use log::trace;
 use std::collections::{HashMap, HashSet};
 
-struct State<'a, 'cfg, 'tmp> {
-    cx: &'tmp mut Context<'a, 'cfg>,
+/// The dependency graph of Units.
+pub type UnitGraph<'a> = HashMap<Unit<'a>, Vec<UnitDep<'a>>>;
+
+/// A unit dependency.
+#[derive(Debug, Clone, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub struct UnitDep<'a> {
+    /// The dependency unit.
+    pub unit: Unit<'a>,
+    /// The purpose of this dependency (a dependency for a test, or a build
+    /// script, etc.).
+    pub unit_for: UnitFor,
+    /// The name the parent uses to refer to this dependency.
+    pub extern_crate_name: InternedString,
+    /// Whether or not this is a public dependency.
+    pub public: bool,
+}
+
+/// Collection of stuff used while creating the `UnitGraph`.
+struct State<'a, 'cfg> {
+    bcx: &'a BuildContext<'a, 'cfg>,
     waiting_on_download: HashSet<PackageId>,
     downloads: Downloads<'a, 'cfg>,
+    unit_dependencies: UnitGraph<'a>,
+    package_cache: HashMap<PackageId, &'a Package>,
+    usr_resolve: &'a Resolve,
+    std_resolve: Option<&'a Resolve>,
+    /// This flag is `true` while generating the dependencies for the standard
+    /// library.
+    is_std: bool,
 }
 
 pub fn build_unit_dependencies<'a, 'cfg>(
-    cx: &mut Context<'a, 'cfg>,
+    bcx: &'a BuildContext<'a, 'cfg>,
+    resolve: &'a Resolve,
+    std_resolve: Option<&'a Resolve>,
     roots: &[Unit<'a>],
-) -> CargoResult<()> {
-    assert!(
-        cx.unit_dependencies.is_empty(),
-        "can only build unit deps once"
-    );
-
+    std_roots: &[Unit<'a>],
+) -> CargoResult<UnitGraph<'a>> {
     let mut state = State {
-        downloads: cx.bcx.packages.enable_download()?,
-        cx,
+        bcx,
+        downloads: bcx.packages.enable_download()?,
         waiting_on_download: HashSet::new(),
+        unit_dependencies: HashMap::new(),
+        package_cache: HashMap::new(),
+        usr_resolve: resolve,
+        std_resolve,
+        is_std: false,
     };
 
+    let std_unit_deps = calc_deps_of_std(&mut state, std_roots)?;
+
+    deps_of_roots(roots, &mut state)?;
+    super::links::validate_links(state.resolve(), &state.unit_dependencies)?;
+    // Hopefully there aren't any links conflicts with the standard library?
```

Comment:
FWIW I think it's safe to say that we should never link in anything with `links` into the standard library, or if we do that we've got bigger fish to fry than Cargo's `links` attribute